compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -2125,25 +2126,43 @@ fn prefetch_mir(tcx: TyCtxt<'_>) {
 // will allow us to slice the metadata to the precise length that we just
 // generated regardless of trailing bytes that end up in it.
 
-#[derive(Encodable, Decodable)]
 pub struct EncodedMetadata {
-    raw_data: Vec<u8>,
+    mmap: Option<Mmap>,
+    decoded: Vec<u8>,
 }
 
 impl EncodedMetadata {
     #[inline]
-    pub fn new() -> EncodedMetadata {
-        EncodedMetadata { raw_data: Vec::new() }
+    pub fn from_file(file: std::fs::File) -> std::io::Result<Self> {
+        let file_metadata = file.metadata()?;
+        if file_metadata.len() == 0 {
+            return Ok(Self { mmap: None, decoded: Vec::new() });
+        }
+        let mmap = unsafe { Some(Mmap::map(file)?) };
+        Ok(Self { mmap, decoded: Vec::new() })
     }
 
     #[inline]
-    pub fn from_raw_data(raw_data: Vec<u8>) -> Self {
-        Self { raw_data }
+    pub fn raw_data(&self) -> &[u8] {
+        if !self.decoded.is_empty() {
+            return &self.decoded;
+        }
+        self.mmap.as_ref().map(|mmap| mmap.as_ref()).unwrap_or_default()
     }
+}
 
-    #[inline]
-    pub fn raw_data(&self) -> &[u8] {
-        &self.raw_data
+impl<S: Encoder> Encodable<S> for EncodedMetadata {
```

Comment:
Are these Encodable/Decodable impls used anywhere?

Reply:
Yes, the metadata is included in `CodegenResults` which can be serialized into a `.rlink` file when using `-Zlink-only`. Might be nice to move the writing of both the rlib metadata object and dylib compressed metadata object to before the linking phase so the `.rlink` file wouldn't need to contain the full metadata and instead just reference the written object files for use during linking.